GUI interface to SHAMPOO

LaurentRDC opened the pull request:
This is a minimal working example.

So far, the only implemented feature is visualization of raw data and reconstructed hologram (amplitude + phase). This pull request is meant as the start of a conversation, as of what to implement next.

bmorris3 committed:
Fixing crop shape bug

LaurentRDC committed:
Initial commit: minimal working example

Minimal working example of the GUI interface to SHAMPOO.

LaurentRDC committed:
Update reconstruction.py

bmorris3 commented on shampoo/gui_utils.py:
```diff
@@ -0,0 +1,141 @@
+"""
+Convenience functions and objects for SHAMPOO's GUI component.
+
+Author: Laurent P. René de Cotret
```
I hate to do this to you @LaurentRDC, but the accent in your name will cause the Python 2.7 builds to fail. Would you mind writing it without the accent for py2.7 compatibility?

LaurentRDC replied:
It had never occurred to me. I will update

bmorris3 replied:
I want to ask how you feel about this – don't be offended if you disagree. I haven't put an `author` tag on each of the files in this repository because this git repository has the contribution history logged within each commit. Would you be offended if I scrubbed this line to be consistent with the rest of the package?

LaurentRDC commented:
Forgot about supporting Python 2.7. Will update soon

LaurentRDC committed:
FIX: relative imports

Fixed relative imports so that the gui can be started from the command
line

LaurentRDC committed:
FIX: python 2.7 compatibility

LaurentRDC committed:
FIX: Dependencies typo

LaurentRDC committed:
FIX: removing dependency on PyQt4

LaurentRDC commented:
Fixed all errors I could, sorry it took a few tries. I don't understand what the Travis build error is, can you help?

bmorris3 committed:
fixing linking

LaurentRDC committed:
Merge pull request #1 from bmorris3/LaurentRDC-master

Let's try it

LaurentRDC commented:
@bmorris3 Thanks for your fix

bmorris3 commented:
No problem! Are you familiar with testing GUIs with pytest? I'm happy to accept PRs for the GUI that aren't covered by tests if that's not an option.

LaurentRDC commented:
It's too simple to have logic to test. I'm just using the Hologram.reconstruct method. I'll include tests when there's more logic involved, very soon.

Laurent

On Fri, Oct 7, 2016 at 5:29 PM -0400, "Brett M. Morris" <notifications@github.com<mailto:notifications@github.com>> wrote:

No problem! Are you familiar with testing GUIs? I'm happy to accept PRs for the GUI that aren't covered by tests if that's not an option.

## 

You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHubhttps://github.com/bmorris3/shampoo/pull/14#issuecomment-252365610, or mute the threadhttps://github.com/notifications/unsubscribe-auth/APdZovCoHC_UF1NM36FhTi_R5otrzQb7ks5qxrmYgaJpZM4KQMdw.

bmorris3 commented:
Great, understood. How should I run this demo locally?

LaurentRDC commented:
No I don't mind at all! My code editor puts it automatically.

LaurentRDC commented:
To run the demo, after installation of pyqtgraph, type:

```
from shampoo.gui import run
run()
```

Edit : typo in pyqtgraph

bmorris3 commented:
This is super cool, awesome start @LaurentRDC! I can't wait to see where this goes from here.

Would you like me to merge this provisionally, so we can distribute it to the Caltech collaborators as a demo? We can add docs and tutorials later as we settle on an interface.

LaurentRDC commented:
Maybe give me a few more days to add some more features? The multiple propagation distances is a big one. I'm not too familiar with GitHub, can you merge this PR in a separate branch? A new 'GUI' branch for example.

bmorris3 commented:
There's no penalty for continuing to commit on this branch (your `master` branch). Let's keep working on this branch, and your future commits will be added to this PR when you `push` them. 

I've had some strange behavior doing reconstructions with this branch via the GUI on my computer. I tried running reconstructions maybe 15 times, and one of those times the reconstruction completed successfully at the right propagation distance, but the other times the propagation distance seemed to have been wrong. Might you know where that bug is coming from?

LaurentRDC commented:
Ok I'll keep pushing to this place.

About your bug, I'll look into it very soon. I'm rewriting the background processes so that it will be easy to extend to real-time inputs. Stay tuned.

LaurentRDC committed:
Reactor-based program structure, easily extensible to real-time processing

The reconstruction of holograms was put in a separate OS process,
allowing for multithreading in the future.

Known issues:
- Cannot set propagation distance

LaurentRDC committed:
Thread-safe communication between reactors

LaurentRDC committed:
Code reorganization: GUI in a subpackage

LaurentRDC committed:
Simplified reactor API

LaurentRDC committed:
Updated documentation

LaurentRDC committed:
Initial camera interface commit

Modified from the pymba library

LaurentRDC committed:
FIX: loading Visual C runtime

LaurentRDC committed:
FIX: Travis CI build errors due to missing dlls

LaurentRDC committed:
Reactor now has a thread-safe stop() method

LaurentRDC committed:
ProcessReactor: Reactor running in a separate process

LaurentRDC committed:
Import fix

LaurentRDC committed:
Import fix

LaurentRDC committed:
Import fix

LaurentRDC committed:
Camera API code cleanup

LaurentRDC committed:
GUI reconstruction runs in a separate core

LaurentRDC committed:
Unnecessary __del__ override

LaurentRDC committed:
Updated documentation

LaurentRDC committed:
Tweaking API calls

LaurentRDC committed:
Tweaking API calls

LaurentRDC committed:
Tested camera frame capture

LaurentRDC committed:
Camera tweaks

LaurentRDC committed:
Quick snapshots

LaurentRDC committed:
API tweaks for compatibility with Python 3

LaurentRDC committed:
Live Viewer for Allied Vision Camera

LaurentRDC committed:
GUI tweaks

LaurentRDC committed:
Enable camera snapshots

LaurentRDC commented:
@bmorris3 

Do you know of a way to stop the continuous integration tools from throwing an error because of missing DLLs? DLLs are loaded as part of a class attribute. Since DLLs are not portable, I can't include it in the repository.

Let me know if you have an idea.

bmorris3 commented:
One thing might be to import the camera module in the top-level `__init__.py` function like the others are ([see here](https://github.com/LaurentRDC/shampoo/blob/master/shampoo/__init__.py#L7-L8)). 

The [problem](https://travis-ci.org/bmorris3/shampoo/jobs/170142783#L726) is caused by [this line](https://github.com/LaurentRDC/shampoo/blob/master/shampoo/camera/allied_vision/vimbacamera.py#L6): 

``` python
from .vimbadll import VimbaDLL
```

 Perhaps you can move that import line into the functions that use that module? That way it won't get called when building the package.

LaurentRDC committed:
VimbaDLL has instance attributes instead of class attributes.

LaurentRDC committed:
Fix to circular import

LaurentRDC committed:
Temporary fix: camera can only output square resolution

LaurentRDC committed:
Merge remote-tracking branch 'refs/remotes/bmorris3/fix-crop-size'

LaurentRDC committed:
Merge remote-tracking branch 'refs/remotes/bmorris3/nonsquareholos'

LaurentRDC committed:
Included PyQtGraph, flattened GUI folder structure

LaurentRDC committed:
Added pyqt4 conda dependency

LaurentRDC committed:
Trying to have a successful build...

LaurentRDC committed:
Remove temporary fix on image size

LaurentRDC committed:
Fixes for API startup

LaurentRDC committed:
Fix for program exit hanging

LaurentRDC committed:
GUI interface to choosing and connecting to a camera

LaurentRDC committed:
Camera snapshots are only available in menu when a camera is connected

LaurentRDC committed:
Better propagation distance selection: start, stop, step

LaurentRDC committed:
debug mode and camera features

LaurentRDC committed:
Graphical interface to camera features (read-only)

LaurentRDC committed:
Merge remote-tracking branch 'refs/remotes/bmorris3/master'

LaurentRDC committed:
Read-write control of some camera features: exposure, bit depth

LaurentRDC committed:
Reconstruction is updated with propagation distance is changed.

LaurentRDC committed:
GUI status bar shows simple actions

Fixes #4

LaurentRDC committed:
Better decoupling of the GUI and logic

LaurentRDC committed:
Better cleanup after quitting

Fixes #5

LaurentRDC committed:
FIX: setting exposure can only be done in increments

LaurentRDC committed:
Testing on allied vision camera

LaurentRDC committed:
Testing on allied vision camera

LaurentRDC committed:
FIX: camera features

LaurentRDC committed:
GUI tweaks

- Prepared saving of raw data

LaurentRDC committed:
Skeleton of HDF5 support

LaurentRDC committed:
Skeleton of viewing what goes on in the Fourier plane

LaurentRDC committed:
Save raw data as TIFF

LaurentRDC committed:
Minor edits

LaurentRDC committed:
FIX: initial propagation distance not updating

LaurentRDC committed:
Fourier plane on log scale

LaurentRDC committed:
FIX: shifted Fourier plane

LaurentRDC committed:
Merge remote-tracking branch 'refs/remotes/bmorris3/master'

LaurentRDC committed:
Merge remote-tracking branch 'refs/remotes/bmorris3/master'

LaurentRDC committed:
Merge remote-tracking branch 'refs/remotes/bmorris3/master'

# Conflicts:
#	.travis.yml
#	appveyor.yml
#	setup.py
#	shampoo/reconstruction.py

LaurentRDC committed:
Python 2.7 compatibility fix

LaurentRDC committed:
Python 2.7 compatibility fix

LaurentRDC committed:
Pip dependencies in Travis and Appveyor

LaurentRDC committed:
FFTutils fix

LaurentRDC committed:
Hologram test fix

LaurentRDC committed:
Docstrings ParseError fix

LaurentRDC committed:
Docstrings ParseError fix

LaurentRDC committed:
Fixes on testing reactors

LaurentRDC committed:
Python 2.7 absolute imports compatibility fix

LaurentRDC committed:
Absolute imports fix

LaurentRDC committed:
Daemon keyword does not exist in Py2.7 Threading module

LaurentRDC committed:
Update to PyQtGraph 0.10 and PyQt 5+

LaurentRDC committed:
Concatenating viewers into QTabWidgets

LaurentRDC committed:
Specialized Reactors to reconstruction only

LaurentRDC committed:
FIX: test_reactor hanging

LaurentRDC committed:
FIX: build_sphinx warnings

LaurentRDC committed:
Automated error messages

LaurentRDC committed:
Enhancement: Hologram reconstruction possible with arbitrary Fourier masks

While the default option is an automatic generation of a Fourier domain
mask, the user can now specify an array to use as a mask. Caching of
hologram reconstructions is temporarily disabled until cache keys can be
updated to take into account the fact that masks can be user-defined.

LaurentRDC committed:
ReconstructedWave has a fourier_mask attribute

LaurentRDC committed:
Removal pyfftw as a required dependency

LaurentRDC committed:
FIX: tests

LaurentRDC committed:
Vizualization of the Fourier mask in GUI

LaurentRDC committed:
Fourier mask loading in GUI is working

LaurentRDC committed:
HDF5 Time series

LaurentRDC committed:
h5py as a dependency

LaurentRDC committed:
Included reconstruction Fourier mask into TimeSeries

LaurentRDC committed:
Refactored Time Series for performance

LaurentRDC committed:
FIX: reconstruction via TimeSeries raises exception

LaurentRDC committed:
IN PROGRESS: Generalizing Hologram class to multiwavelength

LaurentRDC committed:
Generalization to multiwavelength

LaurentRDC committed:
Shampoo's GUI can be called from command line

with the addition of a __main__ module, shampoo's GUI can be opened by:
> python -m shampoo.gui

LaurentRDC committed:
Multiwavelength reconstruction

On the surface, multiwavelength reconstruction is working. The special
case of one wavelength is definitely giving sensical results.

LaurentRDC committed:
FIX: phase unwrapping applied independtly for each wavelength

LaurentRDC committed:
FIX: Python 2 compatibility

LaurentRDC committed:
Time Series Creator in GUI

LaurentRDC committed:
Fixes and test for storage of three-wavelength time series

LaurentRDC committed:
Three-wavelength TimeSeriesCreator

LaurentRDC committed:
Code cleanup

LaurentRDC committed:
Code refactoring

LaurentRDC committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	shampoo/tests/test_time_series.py
#	shampoo/time_series.py

LaurentRDC committed:
Code refactoring

A lot of cleaning up for new features:
- Changed implementation of reconstructor to a QObject
- Added more tests for reconstruction through the GUI
- Fixed reconstruction of three wavelength holograms

LaurentRDC committed:
Refactored TimeSeries structure

More flexible HDF5 file structure allows for more intuitive code

LaurentRDC committed:
Removed real-time reconstruction

LaurentRDC committed:
Reconstruction dialog backbone

LaurentRDC committed:
Time series batch reconstructed tested

zacsimile committed:
Updated reconstruction.py to multiwavelength reconstruction. Errors are still in here, I'm sure.

zacsimile committed:
working three-wavelength reconstruction. no automated spectral peaks or chromatic shifts

zacsimile committed:
Removed print statements

LaurentRDC committed:
Merge pull request #7 from zacsimile/master

Update multiwavelength reconstruction

zacsimile committed:
Merge pull request #1 from LaurentRDC/master

Merge pull request #7 from zacsimile/master

zacsimile committed:
Automatic peak detection. Removed buffer in Fourier mask.

zacsimile committed:
Three-wavelength phase unwrapping.

zacsimile committed:
Updated _find_peak_centroid and centroid tests in test_hologram. All working correctly (I think) now.

LaurentRDC committed:
Merge pull request #8 from zacsimile/master

Added automatic peak centroid finding and three-wavelength phase unwrapping.

LaurentRDC committed:
Tweaks  post-pull request

LaurentRDC committed:
Merge pull request #9 from LaurentRDC/offline-reconstructions-only

Offline reconstructions only

LaurentRDC committed:
Updated docs

LaurentRDC committed:
Time-series support the new three-wavelength reconstruction API

LaurentRDC committed:
FIX: reconstruction dialog

LaurentRDC committed:
GUI refactoring

LaurentRDC committed:
Removed outdated tests

LaurentRDC committed:
Added support for user-provided Fourier mask in reconstruction dialog

LaurentRDC committed:
GUI tweaks

LaurentRDC committed:
Changed format of return value for Hologram.reconstruct for multiple depths

LaurentRDC committed:
Added 'depths' attribute to ReconstructedWave

LaurentRDC committed:
Fixed unwrapped phase dimensions

LaurentRDC committed:
FIX: display axes for reconstructed wave were flipped

LaurentRDC committed:
More robust TimeSeries handling in GUI

Only the ShampooController object can interact with a TimeSeries object.
Hence, it is now extremely unlikely to run into a permission error while
creating or reconstructing a time-series

LaurentRDC committed:
Added chromatic shift to reconstruction dialog

LaurentRDC committed:
Added a way to close time-series

LaurentRDC committed:
Activated continuous integration on Appveyor. Dropped support for python 2.7

LaurentRDC committed:
FIX: chromatic shifts being included when creating time-series through GUI

LaurentRDC committed:
Added GUI metaclass for crash prevention

Instead of crashing, errors will be caught and displayed to the user via
an error message dialog

LaurentRDC committed:
FIX: error message piling up in memory

LaurentRDC committed:
FIX: hologram size

Holograms can only be 2D images. Old code has been updated to check
this.

LaurentRDC committed:
FIX: displaying the log scale fourier intensity

LaurentRDC committed:
Added a way to design Fourier masks
